lib/ProductOpener/Checkpoint.pm

```diff
@@ -0,0 +1,73 @@
+# This file is part of Product Opener.
+#
+# Product Opener
+# Copyright (C) 2011-2025 Association Open Food Facts
+# Contact: contact@openfoodfacts.org
+# Address: 21 rue des Iles, 94100 Saint-Maur des Fossés, France
+#
+# Product Opener is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package ProductOpener::Checkpoint;
+
+use ProductOpener::PerlStandards;
+
+use experimental 'smartmatch';
+use File::Basename qw/basename/;
+use List::MoreUtils qw/first_index/;
+
+use ProductOpener::Paths qw/:all/;
+
+sub new ($class) {
+	my $script_name = basename($0);
+	my $filename = "$BASE_DIRS{CACHE_TMP}/$script_name.checkpoint";
+	if (!-e $filename) {
+		`touch $filename`;
+	}
+	open(my $checkpoint_file, '+<', $filename) or die "Could not open file '$filename' $!";
+	my $checkpoint;
+	my $is_resume = first_index {$_ eq "resume"} @ARGV;
+	if ($is_resume > -1) {
+		seek($checkpoint_file, 0, 0);
+		$checkpoint = <$checkpoint_file>;
+		chomp $checkpoint if $checkpoint;
+		if ($checkpoint) {
+			print STDERR "Resuming from $checkpoint\n";
```

Comment:
Is it assumed that the `Checkpoint` class is always instantiated in a CLI script? Can we check the environment to avoid using it in an unsupported one?

Reply:
I guess we could but is it really a concern? I'd rather not add further complexity if it can be avoided.